sdk/search/azure-search/azure/search/_index/aio/_paging.py

```diff
@@ -0,0 +1,91 @@
+
+from azure.core.async_paging import AsyncItemPaged, AsyncPageIterator, ReturnType
+from .._generated.models import SearchRequest
+from .._paging import (
+    convert_search_result,
+    pack_continuation_token,
+    unpack_continuation_token,
+)
+
+if TYPE_CHECKING:
+    # pylint:disable=unused-import,ungrouped-imports
+    from typing import Any, Union
+
+
+class AsyncSearchItemPaged(AsyncItemPaged[ReturnType]):
+    def __init__(self, *args, **kwargs):
+        super(AsyncSearchItemPaged, self).__init__(*args, **kwargs)
+        self._first_page_iterator_instance = None
+
+    async def __anext__(self) -> ReturnType:
+        if self._page_iterator is None:
+            self._page_iterator = self.by_page()
+            self._first_page_iterator_instance = self._page_iterator
+            return await self.__anext__()
+        if self._page is None:
+            # Let it raise StopAsyncIteration
+            self._page = await self._page_iterator.__anext__()
+            return await self.__anext__()
+        try:
+            return await self._page.__anext__()
+        except StopAsyncIteration:
+            self._page = None
+            return await self.__anext__()
+
+    def _first_iterator_instance(self):
+        if self._first_page_iterator_instance is None:
+            self._page_iterator = self.by_page()
+            self._first_page_iterator_instance = self._page_iterator
+        return self._first_page_iterator_instance
+
+    @property
+    async def facets(self):
+        return await self._first_iterator_instance().facets
+
+
+class AsyncSearchPageIterator(AsyncPageIterator[ReturnType]):
+    def __init__(self, client, initial_query, kwargs, continuation_token=None):
+        super(AsyncSearchPageIterator, self).__init__(
+            get_next=self._get_next_cb,
+            extract_data=self._extract_data_cb,
+            continuation_token=continuation_token,
+        )
+        self._client = client
+        self._initial_query = initial_query
+        self._kwargs = kwargs
+        self._facets = None
+
+    async def _get_next_cb(self, continuation_token):
+        if continuation_token is None:
+            return await self._client.documents.search_post(
+                search_request=self._initial_query.request, **self._kwargs
+            )
+
+        _next_link, next_page_request = unpack_continuation_token(continuation_token)
+
+        return await self._client.documents.search_post(
+            search_request=next_page_request
+        )
+
+    async def _extract_data_cb(self, response):  # pylint:disable=no-self-use
+        continuation_token = pack_continuation_token(response)
+        facets = response.facets
+        if facets is not None:
+            self._facets = {k: [x.as_dict() for x in v] for k, v in facets.items()}
+
+        results = [convert_search_result(r) for r in response.results]
+
+        return continuation_token, results
+
+    @property
+    async def facets(self):
```

Comment:
Having an async property is a bit odd. Should we make this into a function instead? As it is, you might get an http response error on an async property access. Which would be a first for me :).

Reply:
seems reasonable

Reply:
fixed in https://github.com/Azure/azure-sdk-for-python/pull/10182/commits/a961499eaa3ecc3faebd28a047af9d0eada30890